### DISCUSSION TITLE
Benchmark sa.execute vs oracledb. what am i doing wrong?

### DISCUSSION THREAD

**Discussion by WilliamStam** [sqlalchemy/sqlalchemy#10874] [Usage Questions] [answered]
ive recently embraced sqlalchemy for everything, works great. but then i got hit with a 43sec read vs a 3sec and decided to try benchmark it. i think im doing something wrong in this.

https://gist.github.com/WilliamStam/b9bed409e3a754bf05accb95d04bb54e 

being more than double an oracledb.execute() when ran in a 50 loop :/ how can i "debug" to find whats causing it? my thoughts with this benchy was to test the actual execute part. so i kept the connecting part out of the timer and warmed up with the exact same query before starting the loop. so in theory the DB would of cached it. so correct me if im wrong.. this is purely the get records show records benchmark right? 

python 3.12
sqlalchemy: 2.0.25
oracledb: 2.0.1

**Comment by CaselIT** [comment]
Hi,

I haven't had a chance of testing it, but a few comments looking quickly at the code:
- when you do the first query you should consume all the rows too, doing a all / fetchall.
- you are using a session in the async sqlalchemy and a connection on the sync one
- try using `for record in records.all()` in the sqlalchemy, since iter will to `fetchone` for each record and it will be slower than a single `all()` call.

In general you should expect sqlalchemy to be slower than the driver directly, since it's taking the driver result and wrapping it in a `Row`. I can't quantify how much time that is, but I would not expect more than 2 times.

How many rows and how many columns does each row have in your test? (also maybe the types even if it should not came into play with a text query?)
- edit: ok 164 rows, since that's the total shown

**Reply by WilliamStam** [reply]
i was expecting a slow down. but yeah more than like a 5% seems pretty strange. each execute is run about 50 times so the "warmup" i dont think is even really needed (i see the full list times taken and there arent any outliers)

i even tried my own custom Sequence and then for each row generating a dynamic dataclass and adding it to the seq. it added overhead but like 0.01s 

- adding fetchall to the warmup didnt do anything that i can see
- hmm didnt know there was a difference tbh. does sync even have a session? can async work against the connection?
- nopes. no diff

DBMS: Oracle (ver. Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production
Version 19.3.0.0.0)

---------

interestingly tho... this actually seems like an oracledb integration thing. on mysql the results are a lot more favourable. if not downright confusing

```
Sqlalchemy Async
  Records: 245
  Time: 0.8144040107727051
  Average: 0.016288080215454102
-
asyncmy Async
  Records: 245
  Time: 0.8457744121551514
  Average: 0.016915488243103027
-
```
SA is actually faster most times i ran this (small margin but still)


--------------

can anyone confirm the oracle thing?

**Reply by CaselIT** [reply]
One thing to note is that sqlalchemy will create a new cursor each time, it will bot reuse an existing one.

Does changing the oracle code to use a new cursor each time changes things?

> hmm didnt know there was a difference tbh. does sync even have a session? can async work against the connection?

Sure sync has a session. In this case I would not use it, so it's probably better to just replace the session in the async version with the use of a connection (we code would look the same as the sync one, just with await/async keywords

**Reply by WilliamStam** [reply]
even putting the whole connection thing inside the loop for oracledb (both sync and async) still doesnt even come close to SA :( 

think im just loosing my sanity -_- theres no ways these can be legit

in another interesting set of tests... thought maybe cx_oracle vs oracle.. 
https://gist.github.com/WilliamStam/63c1ab8f49018c4526c63a2cd3b19088

----

testing against a production oracle server (municipality) to verify that its not something strange between me and the server. this is a whole different server

```
Sqlalchemy oracledb Sync 
  Records: 188
  Time: 23.188364028930664
  Average: 0.4637672805786133
-
Sqlalchemy cx_oracle sync
  Records: 188
  Time: 24.258026599884033
  Average: 0.48516053199768067
-
cx_oracle Sync
  Records: 188
  Time: 12.20337700843811
  Average: 0.2440675401687622
-
Oracledb Sync
  Records: 188
  Time: 11.967827558517456
  Average: 0.23935655117034912
-
```

**Reply by CaselIT** [reply]
I'll try taking a look. 

>  how many columns does each row have in your test

Could you look into this just so that I can try using a similar setup?

**Comment by zzzeek** [comment]
hi -

we have some good benchmark suites in the examples folder at https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.performance .  you can run these benchmarks against your oracledb directly as these have known "proportions" for performance.

We can try running the "large_resultsets" suite.   For oracle,  there's one little fix we have to make to the scripts (oops), so apply this patch:

```diff
diff --git a/examples/performance/large_resultsets.py b/examples/performance/large_resultsets.py
index 9c0d9fc4e..1f305f2b6 100644
--- a/examples/performance/large_resultsets.py
+++ b/examples/performance/large_resultsets.py
@@ -17,6 +17,7 @@ from sqlalchemy import Column
 from sqlalchemy import create_engine
 from sqlalchemy import Integer
 from sqlalchemy import String
+from sqlalchemy import Identity
 from sqlalchemy.ext.declarative import declarative_base
 from sqlalchemy.orm import Bundle
 from sqlalchemy.orm import Session
@@ -29,7 +30,7 @@ engine = None
 
 class Customer(Base):
     __tablename__ = "customer"
-    id = Column(Integer, primary_key=True)
+    id = Column(Integer, Identity(), primary_key=True)
     name = Column(String(255))
     description = Column(String(255))
```

then we can run this against any number of rows like this, here's 1000 rows:

```
$ python -m examples.performance large_resultsets --num 1000 --dburl oracle+oracledb://scott:tiger@oracle18c/xe
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (1000 iterations); total time 0.060187 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (1000 iterations); total time 0.087401 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (1000 iterations); total time 0.088068 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (1000 iterations); total time 0.087113 sec
test_core_fetchall : Load Core result rows using fetchall. (1000 iterations); total time 0.089579 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (1000 iterations); total time 0.049235 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (1000 iterations); total time 0.048480 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (1000 iterations); total time 0.049858 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (1000 iterations); total time 0.025623 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (1000 iterations); total time 0.026182 sec

```


this is an Oracle database over a home network here.   we can see that for 1000 rows (where the rows are also very small) times are extremely short.

I would suggest using this suite and changing the initial table setup to look more like the table you have in play.   If I had to guess I would say you might have some very large BLOB columns on there or something, or many hundreds of columns.

**Reply by CaselIT** [reply]
even in that log we show a similar trend though mike: 
```
test_core_fetchmany : Load Core result rows using Core / fetchmany. (1000 iterations); total time 0.049858 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (1000 iterations); total time 0.026182 sec
```
about 2x the time.

It's been a while so I don't remember what's the ratio in other backend

**Reply by zzzeek** [reply]
Sure but the poster is showing things like "24 seconds vs 11 seconds for 200 rows", we aren't doing 12 seconds worth of CPU bound work for 200 rows.  that is network overhead of some kind which indicates the two tests are not equivalent functionality.

It looks like we may very well be doing something making oracledb very slow?  The examples here can dump profiles like this:

    python -m examples.performance large_resultsets --dump  --dburl oracle+oracledb://scott:tiger@oracle18c/xe --test test_core_fetchall


Look at the difference for oracledb.cursor.fetchall() for SQLA core:

```
Running setup once...
Tests to run: test_core_fetchall
test_core_fetchall : Load Core result rows using fetchall. (500000 iterations); total fn calls 540943
         540943 function calls (540905 primitive calls) in 16.155 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.485    0.485   16.155   16.155 /home/classic/dev/sqlalchemy/examples/performance/large_resultsets.py:105(test_core_fetchall)
        1    0.003    0.003   15.650   15.650 /home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/result.py:1322(fetchall)
        1    0.253    0.253   15.647   15.647 /home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/result.py:546(_allrows)
        1    0.000    0.000   15.393   15.393 /home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/cursor.py:2102(_fetchall_impl)
        1    0.000    0.000   15.393   15.393 /home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/cursor.py:1129(fetchall)
        1   15.293   15.293   15.393   15.393 /home/classic/dev/python-oracledb/src/oracledb/cursor.py:813(fetchall)
   500037    0.074    0.000    0.074    0.000 {method 'append' of 'list' objects}
    10001    0.019    0.000    0.020    0.000 /usr/lib64/python3.12/inspect.py:371(isfunction)

```

vs. raw DBAPI , note the overall callcounts are very similar, but cursor.fetchall() is twice as fast:

```
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (500000 iterations); total fn calls 520654
         520654 function calls (520625 primitive calls) in 8.038 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    8.038    8.038 /home/classic/dev/sqlalchemy/examples/performance/large_resultsets.py:166(test_dbapi_fetchall_no_object)
        1    0.045    0.045    8.038    8.038 /home/classic/dev/sqlalchemy/examples/performance/large_resultsets.py:173(_test_dbapi_raw)
        1    7.898    7.898    7.973    7.973 /home/classic/dev/python-oracledb/src/oracledb/cursor.py:813(fetchall)
   500033    0.057    0.000    0.057    0.000 {method 'append' of 'list' objects}
     5001    0.015    0.000    0.016    0.000 /usr/lib64/python3.12/inspect.py:371(isfunction)
        1    0.014    0.014    0.014    0.014 /home/classic/dev/python-oracledb/src/oracledb/cursor.py:699(execute)
        1    0.000    0.000    0.004    0.004 /home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py:1513(close)
        1    0.000    0.000    0.004    0.004 /home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py:1390(_checkin)
        1    0.000    0.000    0.004    0.004 /home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py:919(_finalize_fairy)

```


so I'm going to guess this has to be the outputtype handlers we are setting up for rows.

**Reply by CaselIT** [reply]
that may be, but the poster is using a `text()` without any columns so no setinput size there.
Does the result changes if you remove the setinputsizes in the test above?

**Reply by zzzeek** [reply, marked as answer]
I found the problem:

```py
>>> import oracledb
>>> oracledb.defaults.arraysize
100

```

ours:

https://github.com/sqlalchemy/sqlalchemy/blob/a4532ffb58570a682c01b0278a378d138aa6c944/lib/sqlalchemy/dialects/oracle/oracledb.py#L104


set that to 100:

    e = create_engine("oracle+oracledb://...", arraysize=100)


here's the result:

```
test_core_fetchall : Load Core result rows using fetchall. (10000 iterations); total time 0.246341 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (10000 iterations); total time 0.232025 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (10000 iterations); total time 0.239486 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (10000 iterations); total time 0.253768 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (10000 iterations); total time 0.229984 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (10000 iterations); total time 0.219408 sec
```


we should modify the Oracle dialects for "arraysize" to be None and we dont set it if None.

**Reply by CaselIT** [reply]
indeed. is this 2.0 or 2.1? I think we could threat it as a bug and say that it was set to be equal to oracle default

**Reply by CaselIT** [reply]
cx oracle too I would guess has 100

**Reply by CaselIT** [reply]
@zzzeek created https://github.com/sqlalchemy/sqlalchemy/issues/10877

**Reply by zzzeek** [reply]
I think we do it right now for 2.0 and 2.1.   this setting was probably never necessary.    we do read the arraysize in some cases so we'd have to make sure it's present on cursor and such if we didnt actually set it (tests will fail badly if it's not there so we'll know).

**Reply by CaselIT** [reply]
it's documented at 50 so it's probably more of a 2.1 change

**Reply by CaselIT** [reply]
but I'm not opposed at doing this at 2.0, maybe only for oracledb

**Reply by zzzeek** [reply]
OK.  it's too bad because nobody knows about this parameter, I forgot all about it here myself

**Reply by CaselIT** [reply]
I mean I don't think it will cause any issue to anyone, considering that 100 is the default set by oracle

**Reply by zzzeek** [reply]
it's a silly parameter because it's very much the kind of thing that should adjust dynamically per query.   I think we have an execution option for it but IMO the driver should be determining buffering for results on the fly based on what's being requested and delivered

**Reply by WilliamStam** [reply]
i stepped out for a few hours and come back to this... _jaw drops_ 

> How many rows and how many columns does each row have in your test?

for what its worth.. the table: (im just using text to take out the orm factor)

```
class SourceTable(SourceBase):
    __tablename__ = "UPD_AUDT"

    UPD_SEQ: Mapped[int] = mapped_column(Integer, primary_key=True)
    UPD_OPR: Mapped[str] = mapped_column(String(250), nullable=True)
    UPD_DATE: Mapped[datetime.date] = mapped_column(Date, nullable=True)
    UPD_TIME: Mapped[str] = mapped_column(String(10), nullable=True)
    UPD_VERS1: Mapped[str] = mapped_column(String(2), nullable=True)
    UPD_VERS2: Mapped[str] = mapped_column(String(2), nullable=True)
    UPD_VERS3: Mapped[str] = mapped_column(String(2), nullable=True)
    UPD_SUB_VERS: Mapped[str] = mapped_column(String(2), nullable=True)
    UPD_RELEASE_PDF: Mapped[str] = mapped_column(Text, nullable=True)
```
table data is 


but i see the same ratios for the other queries as well. 

> 24 seconds vs 11 seconds for 200 rows

i run the query 50 times in a loop. 

i went drinking (so slightly not able to run the tests now) cause strangely the mysql benchy SA was actually faster than asyncmy (by 0.001s so really nothing much but it was still almost always quicker in SA). to make matters worse i did the same thing between cx_oracle + oracledb + sqlalchemy[cx_oracle] + sqlalchemy[oracledb]

thanks for looking into it on a friday!

**Comment by WilliamStam** [comment]
actually does seem that this is "normal". again, thanks for looking into it. that test suite helped a LOT (honestly didnt know about them). the sqlalchemy tests are indeed about 2 to 2.5 times slower than dbapi.

im pretty sure these mysql tests are flawed...

```
DBMS: MySQL (ver. 8.0.29-21)
Ping: 47 ms

(venv-3.12) W:\Projects\Testing>python -m testing  --dburl mysql+pymysql://username:password@ip/benchmark --num 100000 large_resultsets 
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 1.208659 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 2.155716 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 0.629959 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 0.449096 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 0.850662 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 0.588790 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 0.827598 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 0.655055 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (100000 iterations); total time 0.914830 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 0.674674 sec
``` 

```
DBMS: Oracle (ver. Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - ProductionVersion 19.21.0.0.0)
Ping: 52 ms

(venv-3.12) W:\Projects\Project>python -m testing  --dburl oracle+oracledb://username:password@ip/benchmark --num 100000 large_resultsets
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 217.393872 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 215.886381 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 221.100777 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 215.476969 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 215.640618 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 220.939874 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 194.137611 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 147.895902 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (100000 iterations); total time 56.417417 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 58.702589 sec
```

hardware is very very similar, in the same rack. 

did the same with postgres

```
DBMS: PostgreSQL (ver. 14.10 (Ubuntu 14.10-0ubuntu0.22.04.1))
Ping: 26 ms
(venv-3.12) W:\Projects\Testing>python -m testing  --dburl postgresql+psycopg://username:password@ip/benchmark --num 100000 large_resultsets 
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 1.532622 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 2.826966 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 1.656755 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 1.794711 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 1.257559 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 1.341054 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 1.885715 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 1.532917 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (100000 iterations); total time 1.391414 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 1.399691 sec
```
(worstcase / bestcase)

oracle:
DBAPI:  58.702589 sec
Load fully tracked ORM (yield): 215.886381 sec

postgres:
DBAPI:  1.399691 sec
Load fully tracked ORM (yield): 2.826966 sec
 
mysql:
DBAPI: 2.155716 sec
Load fully tracked ORM (yield): 0.674674 sec

i even tried running the oracle test against 4 production financial servers to make sure its not "this" db as well as a pre prod server and a QA server. it looks the same everywhere. oracle is dog dog dog slow :(

Any reason why adding in objects makes it exponentially slower like that? my thinking here is that you have the record. x. convert it to obj. no matter the server it should still take a row and object it up. why is the "turning it to an object" slower for oracle vs say mysql? still same object?

There is no ways the mysql / postgres can return 100k records in 2sec tho. so something is wrong somewhere :( all i changed in all cases was the --dburl (cx_oracle does the same as oracledb) 

Would this apply?
> Drivers such as those of PostgreSQL and MySQL/MariaDB generally use client side cursors by default. A server side cursor, by contrast, indicates that result rows remain pending within the database server’s state as result rows are consumed by the client. 

please can someone test these results :(

**Reply by zzzeek** [reply]
> mysql:
> DBAPI: 2.155716 sec

I dont see that in your listing?  that timestamp refers to:

    test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 2.155716 sec

that's ORM + yield_per (server side cursor), I expect performance to be poor for that case (mysql server side cursor is not a robust API), plus you have ORM overhead, not unusual

comparing any ORM case to Core / DBAPI is only useful in terms of SQLAlchemy CPU overhead, if you are looking for driver issues you would ignore any line with ORM in it and just look at "Core".  the Oracle result you showed still seems to have something wrong:

     test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 215.640618 sec
     test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 58.702589 sec

did you set arraysize to 100 for that run?

**Reply by WilliamStam** [reply]
> dont see that in your listing? that timestamp refers to:

i got the 2 figures to labels mixed up :( ment to be:
```
mysql:
DBAPI: 0.674674 sec
Load fully tracked ORM (yield): 2.155716 sec
```
 (in the results the orm line is above the dbapi)

> did you set arraysize to 100 for that run?

woops. i changed it in my main apps create_async_engine but not in the test *blush*. 

test suite changed to:

```

@Profiler.setup_once
def setup_database(dburl, echo, num):
    global engine
    engine = create_engine(dburl, echo=echo, arraysize=100) # <-- here
    Base.metadata.drop_all(engine)
    Base.metadata.create_all(engine)
```


----
i bugged the DBA to check the server(s) and he is now giving me lip about wasting his time since everything seems fine in his realm.... 

made a new project. pip install sqlalchemy and oracledb only. copied the test files in and ran them (incase me adding in all the mysql/postgres/cx_oracle messed something up). 


wow... arraysize=100

```
(venv) W:\Projects\test_oracledb>python -m testing large_resultsets --num 100000  --dburl oracle+oracledb://xx:xx@xx/munwh
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 58.255108 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 56.626920 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 60.764838 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 57.262886 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 56.251809 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 55.528010 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 9.965752 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 9.964860 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (100000 iterations); total time 10.046592 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 10.081413 sec
```

arraysize removed, think we found the culprit... 

```
(venv) W:\Projects\test_oracledb>python -m testing large_resultsets --num 100000  --dburl oracle+oracledb://xx:xx@xx/munwh
Running setup once...
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 206.994569 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 209.742395 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 203.139260 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 205.794479 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 207.417368 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 202.066496 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 209.078167 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 202.382260 sec
test_dbapi_fetchall_plus_append_objects : Load rows using DBAPI fetchall(), generate an object for each row. (100000 iterations); total time 54.787318 sec
test_dbapi_fetchall_no_object : Load rows using DBAPI fetchall(), don't make any objects. (100000 iterations); total time 59.178724 sec
```

**Reply by WilliamStam** [reply]
think ive wasted everyone's time here :( sorry!  https://github.com/sqlalchemy/sqlalchemy/issues/10877

**Reply by CaselIT** [reply]
no worries, you prompted a revisitation of a very old assumption in sqlalchemy that hopefully improves things for other people too.

Still it's surprising how effected oracle is by that parameter, and it's surprising that a more automatic / dynamic solution is not implemented in the driver.

**Reply by WilliamStam** [reply]
```
extra = {}
if self.config.driver.startswith("oracle"):
    extra['arraysize'] = 100 # the other servers complain about having it in the create_engine options. 
engine = create_async_engine(
            URL.create(
                drivername=self.config.driver or "sqlite",
                username=self.config.username or None,
                password=self.config.password or None,
                host=self.config.host or None,
                port=self.config.port or None,
                database=self.config.database or None,
            ),
            echo=self.echo,
            **extra
        )
```
with `arraysize` set to `None`

```
(venv) W:\Projects\test_oracledb>python -m testing large_resultsets --num 100000  --dburl oracle+oracledb://xx:xx@xx-poc/xx
Running setup once...
Tests to run: test_orm_full_objects_list, test_orm_full_objects_chunks, test_orm_bundles, test_orm_columns, test_core_fetchall, test_core_fetchall_mapping, test_core_fetchmany_w_streaming, test_core_fetchmany, test_dbapi_fetchall_plus_append_objects, test_dbapi_fetchall_no_object
test_orm_full_objects_list : Load fully tracked ORM objects into one big list(). (100000 iterations); total time 59.923891 sec
test_orm_full_objects_chunks : Load fully tracked ORM objects a chunk at a time using yield_per(). (100000 iterations); total time 59.233323 sec
test_orm_bundles : Load lightweight "bundle" objects using the ORM. (100000 iterations); total time 59.527708 sec
test_orm_columns : Load individual columns into named tuples using the ORM. (100000 iterations); total time 60.097307 sec
test_core_fetchall : Load Core result rows using fetchall. (100000 iterations); total time 61.273475 sec
test_core_fetchall_mapping : Load Core result rows using fetchall. (100000 iterations); total time 59.261249 sec
test_core_fetchmany_w_streaming : Load Core result rows using fetchmany/streaming. (100000 iterations); total time 60.376520 sec
test_core_fetchmany : Load Core result rows using Core / fetchmany. (100000 iterations); total time 60.663039 sec
```

**Reply by WilliamStam** [reply]
just a word of caution. setting arraysize to `None` is better than with it not set, but with 100 it works better

**Reply by CaselIT** [reply]
that's strange since the driver default is 100, so setting it to None or 100 in sqlalchemy should be equal from the driver point of view.

Thanks for looking into it, I'll try doing a few tests on this
